File: app/src/main/java/com/voipgrid/vialer/fcm/FcmMessagingService.java
package com.voipgrid.vialer.fcm;



import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.PowerManager;
import android.telephony.TelephonyManager;
import android.util.Log;

import com.google.firebase.messaging.FirebaseMessagingService;
import com.google.firebase.messaging.RemoteMessage;
import com.voipgrid.vialer.User;
import com.voipgrid.vialer.api.Middleware;
import com.voipgrid.vialer.api.ServiceGenerator;
import com.voipgrid.vialer.call.NativeCallManager;
import com.voipgrid.vialer.logging.LogHelper;
import com.voipgrid.vialer.logging.Logger;
import com.voipgrid.vialer.middleware.MiddlewareHelper;
import com.voipgrid.vialer.notifications.VoipDisabledNotification;
import com.voipgrid.vialer.sip.SipConstants;
import com.voipgrid.vialer.sip.SipService;
import com.voipgrid.vialer.sip.SipUri;
import com.voipgrid.vialer.sip.core.Action;
import com.voipgrid.vialer.sip.service.SipServiceExtra;
import com.voipgrid.vialer.util.ConnectivityHelper;
import com.voipgrid.vialer.util.PhoneNumberUtils;

import androidx.annotation.NonNull;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * Listen to messages from FCM. The backend server sends us FCM notifications when we have
 * incoming calls.
 */
public class FcmMessagingService extends FirebaseMessagingService {

    /**
     * The number of times the middleware will attempt to send a push notification
     * before it gives up.
     *
     */
    private static final int MAX_MIDDLEWARE_PUSH_ATTEMPTS = 8;

    /**
     * Stores the last call we have SUCCESSFULLY handled and started the SipService
     * for.
     */
    private static String sLastHandledCall;

    public static final String VOIP_HAS_BEEN_DISABLED = "com.voipgrid.vialer.voip_disabled";

    private Logger mRemoteLogger;
    private ConnectivityHelper mConnectivityHelper;
    private PowerManager mPowerManager;
    private NativeCallManager nativeCallManager;

    @Override
    public void onCreate() {
        super.onCreate();
        mRemoteLogger = new Logger(FcmMessagingService.class);
        mConnectivityHelper = ConnectivityHelper.get(this);
        mPowerManager = (PowerManager) getSystemService(POWER_SERVICE);
        nativeCallManager = new NativeCallManager((TelephonyManager) getSystemService(Context.TELEPHONY_SERVICE));
        mRemoteLogger.d("onCreate");
    }

    @Override
    public void onMessageReceived(RemoteMessage remoteMessage) {
        super.onMessageReceived(remoteMessage);
        mRemoteLogger.d("onMessageReceived");
        RemoteMessageData remoteMessageData = new RemoteMessageData(remoteMessage.getData());
        LogHelper.using(mRemoteLogger).logMiddlewareMessageReceived(remoteMessage, remoteMessageData.getRequestType());

        if (!remoteMessageData.hasRequestType()) {
            mRemoteLogger.e("No requestType");
            return;
        }

        if (remoteMessageData.isCallRequest()) {
            handleCall(remoteMessage, remoteMessageData);
            return;
        }

        if (remoteMessageData.isMessageRequest()) {
            handleMessage(remoteMessage, remoteMessageData);
            return;
        }
    }

    @Override
    public void onDeletedMessages() {
        super.onDeletedMessages();
        mRemoteLogger.d("Message deleted on the FCM server.");
    }

    /**
     * Handle a push message with a call request type.
     *
     * @param remoteMessage
     * @param remoteMessageData
     */
    private void handleCall(RemoteMessage remoteMessage, RemoteMessageData remoteMessageData) {
        logCurrentState(remoteMessageData);

        if (!isConnectionSufficient()) {
            handleInsufficientConnection(remoteMessage, remoteMessageData);
            return;
        }

        if (isAVialerCallAlreadyInProgress()) {
            rejectDueToVialerCallAlreadyInProgress(remoteMessage, remoteMessageData);
            return;
        }

        if (nativeCallManager.isBusyWithNativeCall()) {
            rejectDueToNativeCallAlreadyInProgress(remoteMessage, remoteMessageData);
            return;
        }

        sLastHandledCall = remoteMessageData.getRequestToken();

        mRemoteLogger.d("Payload processed, calling startService method");

        startSipService(remoteMessageData);
    }

    /**
     * Handle a push message with a message request type.
     *
     * @param remoteMessage
     * @param remoteMessageData
     */
    private void handleMessage(RemoteMessage remoteMessage, RemoteMessageData remoteMessageData) {
        if (!remoteMessageData.isRegisteredOnOtherDeviceMessage()) {
            return;
        }

        if (User.voip.getHasEnabledSip()) {
            new VoipDisabledNotification().display();
        }

        User.voip.setHasEnabledSip(false);
        LocalBroadcastManager.getInstance(this).sendBroadcast(new Intent(VOIP_HAS_BEEN_DISABLED));
    }

    /**
     * Performs various tasks that are required when we are rejecting a call due to an insufficient
     * network connection.
     *
     * @param remoteMessage The remote message that we are handling.
     * @param remoteMessageData The remote message data that we are handling.
     */
    private void handleInsufficientConnection(RemoteMessage remoteMessage, RemoteMessageData remoteMessageData) {
        if (isDeviceInIdleMode()) {
            mRemoteLogger.e("Device in idle mode and connection insufficient. For now do nothing wait for next middleware push.");
        }
        else {
            mRemoteLogger.e("Connection is insufficient. For now do nothing and wait for next middleware push");
        }
    }

    /**
     * Check if we have a good enough connection to accept an incoming call.
     *
     * @return TRUE if we have a good enough connection, otherwise FALSE.
     */
    private boolean isConnectionSufficient() {
        return mConnectivityHelper.hasNetworkConnection() && mConnectivityHelper.hasFastData();
    }

    /**
     * Check to see if the SIP service is currently running, this means that there is already a call
     * in progress and we can not accept further calls.
     *
     * @return TRUE if there is an active call, otherwise FALSE
     */
    private boolean isAVialerCallAlreadyInProgress() {
        return SipService.Companion.getSipServiceActive();
    }

    /**
     * Check if we have reached or exceeded the maximum number of attempts that we
     * accept from the middleware.
     *
     * @param remoteMessageData The remote message data that we are handling.
     * @return TRUE if we have reached or exceeded maximum attempts, otherwise FALSE.
     */
    private boolean hasExceededMaximumAttempts(RemoteMessageData remoteMessageData) {
        return remoteMessageData.getAttemptNumber() >= MAX_MIDDLEWARE_PUSH_ATTEMPTS;
    }

    /**
     * Performs various tasks that are necessary when rejecting a call based on the fact that there is
     * already a Vialer call in progress.
     *
     * @param remoteMessage The remote message that we are handling.
     * @param remoteMessageData The remote message data that we are handling.
     */
    private void rejectDueToVialerCallAlreadyInProgress(RemoteMessage remoteMessage, RemoteMessageData remoteMessageData) {
        mRemoteLogger.d("Reject due to call already in progress");

        rejectCall(remoteMessageData);

        sendCallFailedDueToOngoingVialerCallMetric(remoteMessage, remoteMessageData.getRequestToken());
    }

    /**
     * Performs various tasks that are necessary when rejecting a call based on the fact that there is
     * already a Vialer call in progress.
     *
     * @param remoteMessage The remote message that we are handling.
     * @param remoteMessageData The remote message data that we are handling.
     */
    private void rejectDueToNativeCallAlreadyInProgress(RemoteMessage remoteMessage, RemoteMessageData remoteMessageData) {
        mRemoteLogger.d("Reject due to native call already in progress");

        rejectCall(remoteMessageData);
    }

    /**
     * Send the vialer metric for ongoing call if appropriate.
     *
     * @param remoteMessage
     * @param requestToken
     */
    private void sendCallFailedDueToOngoingVialerCallMetric(RemoteMessage remoteMessage, String requestToken) {
        if (sLastHandledCall != null && sLastHandledCall.equals(requestToken)) {
            mRemoteLogger.i("Push notification (" + sLastHandledCall + ") is being rejected because there is a Vialer call already in progress but not sending metric because it was already handled successfully");
            return;
        }
    }

    /**
     * Notify the middleware server that we are, in fact, alive.
     *
     * @param remoteMessageData The remote message data from the remote message that we are handling.
     */
    private void rejectCall(RemoteMessageData remoteMessageData) {
        if (remoteMessageData.getRequestToken().equals(sLastHandledCall)) return;

        mRemoteLogger.d("replyServer");
        ServiceGenerator.createRegistrationService(this)
                .reply(remoteMessageData.getRequestToken(), false, remoteMessageData.getMessageStartTime())
                .enqueue(null);
    }

    /**
     * Start the SIP service with the relevant data from the push message in the
     * intent.
     *
     * @param remoteMessageData
     */
    private void startSipService(RemoteMessageData remoteMessageData) {
        mRemoteLogger.d("startSipService");
        Intent intent = new Intent(this, SipService.class);
        intent.setAction(Action.HANDLE_INCOMING_CALL.toString());

        // Set a phoneNumberUri as DATA for the intent to SipServiceOld.
        Uri sipAddressUri = SipUri.sipAddressUri(
                this,
                PhoneNumberUtils.format(remoteMessageData.getPhoneNumber())
        );

        intent.setData(sipAddressUri);

        intent.putExtra(SipServiceExtra.INCOMING_TOKEN.name(), remoteMessageData.getRequestToken());
        intent.putExtra(SipServiceExtra.INCOMING_CALL_START_TIME.name(), remoteMessageData.getMessageStartTime());

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            startForegroundService(intent);
        } else {
            startService(intent);
        }
    }

    /**
     * Device can ben in Idle mode when it's been idling to long. This means that network connectivity
     * is reduced. So we check if we are in that mode and the connection is insufficient.
     * just return and don't reply to the middleware for now.
     *
     * @return
     */
    private boolean isDeviceInIdleMode() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && mPowerManager.isDeviceIdleMode();
    }

    /**
     * Log some information about our current state to help determine what state the phone is in when
     * a push notification is incoming.
     *
     * @param remoteMessageData
     */
    private void logCurrentState(RemoteMessageData remoteMessageData) {
        mRemoteLogger.d("SipService Active: " + SipService.Companion.getSipServiceActive());
        mRemoteLogger.d("CurrentConnection: " + mConnectivityHelper.getConnectionTypeString());
        mRemoteLogger.d("Payload: " + remoteMessageData.getRawData().toString());
    }

    @Override
    public void onNewToken(final String s) {
        super.onNewToken(s);
        mRemoteLogger.d("onTokenRefresh");
        MiddlewareHelper.setRegistrationStatus(
                com.voipgrid.vialer.persistence.Middleware.RegistrationStatus.UNREGISTERED);
        MiddlewareHelper.registerAtMiddleware(this);
    }
}

File: app/src/main/java/com/voipgrid/vialer/fcm/FcmMessagingService.kt
package com.voipgrid.vialer.fcm

import android.content.Intent
import android.os.Build
import android.os.PowerManager
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.google.firebase.messaging.FirebaseMessagingService
import com.google.firebase.messaging.RemoteMessage
import com.voipgrid.vialer.User
import com.voipgrid.vialer.api.Middleware
import com.voipgrid.vialer.call.NativeCallManager
import com.voipgrid.vialer.logging.LogHelper
import com.voipgrid.vialer.logging.Logger
import com.voipgrid.vialer.notifications.VoipDisabledNotification
import com.voipgrid.vialer.phonelib.Initialiser
import com.voipgrid.vialer.sip.SipConstants
import com.voipgrid.vialer.sip.SipService
import com.voipgrid.vialer.sip.SipUri
import com.voipgrid.vialer.statistics.VialerStatistics
import com.voipgrid.vialer.util.ConnectivityHelper
import com.voipgrid.vialer.util.PhoneNumberUtils
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import org.koin.android.ext.android.inject
import org.koin.core.component.KoinComponent

/**
 * Listen to messages from FCM. The backend server sends us FCM notifications when we have
 * incoming calls.
 */
class FcmMessagingService : FirebaseMessagingService(), KoinComponent {

    private val logger = Logger(this)
    private val connectivityHelper: ConnectivityHelper by inject()
    private val powerManager: PowerManager by inject()
    private val nativeCallManager: NativeCallManager by inject()
    private val middlewareApi: Middleware by inject()
    private val middleware: com.voipgrid.vialer.middleware.Middleware by inject()
    private val initialiser: Initialiser by inject()

    /**
     * The number of times the middleware will attempt to send a push notification
     * before it gives up and the string stores the last call we have SUCCESSFULLY handled and
     * started the SipService for.
     *
     */
    companion object {
        private const val MAX_MIDDLEWARE_PUSH_ATTEMPTS = 8
        private var lastHandledCall: String? = null
        const val VOIP_HAS_BEEN_DISABLED = "com.voipgrid.vialer.voip_disabled"
    }

    override fun onMessageReceived(remoteMessage: RemoteMessage) {
        super.onMessageReceived(remoteMessage)
        val remoteMessageData = RemoteMessageData(remoteMessage.data)
        LogHelper.using(logger).logMiddlewareMessageReceived(remoteMessage, remoteMessageData.requestType)
        VialerStatistics.pushNotificationWasReceived(remoteMessage)

        when {
            !remoteMessageData.hasRequestType() -> logger.e("No requestType")
            remoteMessageData.isCallRequest -> handleCall(remoteMessage, remoteMessageData)
            remoteMessageData.isMessageRequest -> handleMessage(remoteMessage, remoteMessageData)
        }
    }

    override fun onDeletedMessages() {
        super.onDeletedMessages()
        logger.d("Message deleted on the FCM server.")
    }

    /**
     * Handle a push message with a call request type.
     *
     * @param remoteMessage
     * @param remoteMessageData
     */
    private fun handleCall(remoteMessage: RemoteMessage, remoteMessageData: RemoteMessageData) {
        logCurrentState(remoteMessageData)

        when {
            !isConnectionSufficient() -> handleInsufficientConnection(remoteMessage, remoteMessageData)
            isAVialerCallAlreadyInProgress() -> rejectDueToVialerCallAlreadyInProgress(remoteMessage, remoteMessageData)
            nativeCallManager.isBusyWithNativeCall ->  rejectDueToNativeCallAlreadyInProgress(remoteMessage, remoteMessageData)
            else -> {


                if (lastHandledCall != remoteMessageData.requestToken) {
                    logger.d("Payload processed, calling startService method")
                    lastHandledCall = remoteMessageData.requestToken
                    startSipService(remoteMessageData)
                }
            }
        }
    }

    /**
     * Handle a push message with a message request type.
     *
     * @param remoteMessage
     * @param remoteMessageData
     */
    private fun handleMessage(remoteMessage: RemoteMessage, remoteMessageData: RemoteMessageData) {

        if (!remoteMessageData.isRegisteredOnOtherDeviceMessage) {
            return
        }

        if (User.voip.hasEnabledSip) {
            VoipDisabledNotification().display()
            User.voip.hasEnabledSip = false
        }

        LocalBroadcastManager.getInstance(this).sendBroadcast(Intent(VOIP_HAS_BEEN_DISABLED))
    }

    /**
     * Performs various tasks that are required when we are rejecting a call due to an insufficient
     * network connection.
     *
     * @param remoteMessage The remote message that we are handling.
     * @param remoteMessageData The remote message data that we are handling.
     */
    private fun handleInsufficientConnection(remoteMessage: RemoteMessage, remoteMessageData: RemoteMessageData) {
        if (hasExceededMaximumAttempts(remoteMessageData)) {
            VialerStatistics.incomingCallFailedDueToInsufficientNetwork(remoteMessage)
        }
        logger.e(when(isDeviceInIdleMode()){
            true -> "Device in idle mode and connection insufficient. For now do nothing wait for next middleware push."
            false -> "Connection is insufficient. For now do nothing and wait for next middleware push"
        })
    }

    /**
     * Check if we have a good enough connection to accept an incoming call.
     *
     * @return TRUE if we have a good enough connection, otherwise FALSE.
     */
    private fun isConnectionSufficient() = connectivityHelper.hasNetworkConnection() && connectivityHelper.hasFastData()

    /**
     * Check to see if the SIP service is currently running, this means that there is already a call
     * in progress and we can not accept further calls.
     *
     * @return TRUE if there is an active call, otherwise FALSE
     */
    private fun isAVialerCallAlreadyInProgress() = SipService.sipServiceActive;

    /**
     * Check if we have reached or exceeded the maximum number of attempts that we
     * accept from the middleware.
     *
     * @param remoteMessageData The remote message data that we are handling.
     * @return TRUE if we have reached or exceeded maximum attempts, otherwise FALSE.
     */
    private fun hasExceededMaximumAttempts(remoteMessageData: RemoteMessageData) = remoteMessageData.getAttemptNumber() >= MAX_MIDDLEWARE_PUSH_ATTEMPTS

    /**
     * Performs various tasks that are necessary when rejecting a call based on the fact that there is
     * already a Vialer call in progress.
     *
     * @param remoteMessage The remote message that we are handling.
     * @param remoteMessageData The remote message data that we are handling.
     */
    private fun rejectDueToVialerCallAlreadyInProgress(remoteMessage: RemoteMessage, remoteMessageData: RemoteMessageData) {
        logger.d("Reject due to call already in progress")

        replyServer(remoteMessageData, false)

        sendCallFailedDueToOngoingVialerCallMetric(remoteMessage, remoteMessageData.requestToken)
    }

    /**
     * Performs various tasks that are necessary when rejecting a call based on the fact that there is
     * already a Vialer call in progress.
     *
     * @param remoteMessage The remote message that we are handling.
     * @param remoteMessageData The remote message data that we are handling.
     */
    private fun rejectDueToNativeCallAlreadyInProgress(remoteMessage: RemoteMessage, remoteMessageData: RemoteMessageData) {
        logger.d("Reject due to native call already in progress")

        replyServer(remoteMessageData, false)

        VialerStatistics.incomingCallFailedDueToOngoingGsmCall(remoteMessage)
    }

    /**
     * Send the vialer metric for ongoing call if appropriate.
     *
     * @param remoteMessage
     * @param requestToken
     */
    private fun sendCallFailedDueToOngoingVialerCallMetric(remoteMessage: RemoteMessage, requestToken: String) {
        if (lastHandledCall != null && lastHandledCall == requestToken) {
            logger.i("Push notification ($lastHandledCall) is being rejected because there is a Vialer call already in progress but not sending metric because it was already handled successfully")
            return
        }

        VialerStatistics.incomingCallFailedDueToOngoingVialerCall(remoteMessage)
    }

    /**
     * Notify the middleware server that we are, in fact, alive.
     *
     * @param remoteMessageData The remote message data from the remote message that we are handling.
     * @param isAvailable TRUE if the phone is ready to accept the incoming call, FALSE if it is not available.
     */
    private fun replyServer(remoteMessageData: RemoteMessageData, isAvailable: Boolean) = GlobalScope.launch {
        val response =  middlewareApi.reply(
                remoteMessageData.requestToken,
                isAvailable,
                remoteMessageData.messageStartTime,
                User.voipAccount?.accountId
        ).execute()
        if (response.isSuccessful) {
            logger.i("response was successful")
        }
    }

    /**
     * Start the SIP service with the relevant data from the push message in the
     * intent.
     *
     * @param remoteMessageData
     */
    private fun startSipService(remoteMessageData: RemoteMessageData) {
        val intent = Intent(this, SipService::class.java).apply {
            action = SipService.Actions.HANDLE_INCOMING_CALL
            data = SipUri.sipAddressUri(this@FcmMessagingService,PhoneNumberUtils.format(remoteMessageData.phoneNumber))
            putExtra(SipConstants.EXTRA_RESPONSE_URL, remoteMessageData.responseUrl);
            putExtra(SipConstants.EXTRA_REQUEST_TOKEN, remoteMessageData.requestToken)
            putExtra(SipConstants.EXTRA_PHONE_NUMBER, remoteMessageData.phoneNumber)
            putExtra(SipConstants.EXTRA_CONTACT_NAME, remoteMessageData.callerId)
            putExtra(RemoteMessageData.MESSAGE_START_TIME, remoteMessageData.messageStartTime)
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            startForegroundService(intent)
        } else {
            startService(intent)
        }
    }

    /**
     * Device can be in Idle mode when it's been idling to long. This means that network connectivity
     * is reduced. So we check if we are in that mode and the connection is insufficient.
     * Just return and don't reply to the middleware for now.
     *
     * @return
     */
    private fun isDeviceInIdleMode() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && powerManager.isDeviceIdleMode

    /**
     * Log some information about our current state to help determine what state the phone is in when
     * a push notification is incoming.
     *
     * @param remoteMessageData
     */
    private fun logCurrentState(remoteMessageData: RemoteMessageData) {
        listOf(
                "SipService Active: " + SipService.sipServiceActive,    "CurrentConnection: " + connectivityHelper.getConnectionTypeString(),
                "Payload: " + remoteMessageData.getRawData().toString()
        ).forEach{
            logger.d(it)
        }
    }

    override fun onNewToken(s: String) {
        super.onNewToken(s)
        middleware.register()
    }
}

